pkg/relayer/miner/miner.go

```diff
@@ -0,0 +1,395 @@
+}
+
+// waitForEarliestCreateClaimHeight returns the earliest block height at which a
+// claim can be created. It is calculated from the session end block height,
+// on-chain governance parameters, and randomized input.
+func (mnr *miner) waitForEarliestCreateClaimHeight(
+	ctx context.Context,
+	sessionEndHeight int64,
+) {
+	// TODO_TECHDEBT: refactor this logic to a shared package.
+
+	createClaimWindowStartHeight := sessionEndHeight
+	// TODO_TECHDEBT: query the on-chain governance parameter once available.
+	// + claimproofparams.GovEarliestClaimSubmissionBlocksOffset
+
+	// we wait for createClaimWindowStartHeight to be received before proceeding since we need its hash
+	// to know where this servicer's claim submission window starts.
+	log.Printf("waiting for global earliest claim submission createClaimWindowStartBlock height: %d", createClaimWindowStartHeight)
+	createClaimWindowStartBlock := mnr.waitForBlock(ctx, createClaimWindowStartHeight)
+
+	log.Printf("received earliest claim submission createClaimWindowStartBlock height: %d, use its hash to have a random submission for the servicer", createClaimWindowStartBlock.Height())
+
+	earliestCreateClaimHeight :=
+		protocol.GetEarliestCreateClaimHeight(createClaimWindowStartBlock)
+
+	log.Printf("earliest claim submission createClaimWindowStartBlock height for this supplier: %d", earliestCreateClaimHeight)
+	_ = mnr.waitForBlock(ctx, earliestCreateClaimHeight)
+}
+
+// waitForBlock blocks until the block at the given height (or greater) is
+// observed as having been committed.
+func (mnr *miner) waitForBlock(ctx context.Context, height int64) client.Block {
+	subscription := mnr.blockClient.CommittedBlocksSequence(ctx).Subscribe(ctx)
+	defer subscription.Unsubscribe()
+
+	for block := range subscription.Ch() {
+		if block.Height() >= height {
+			return block
+		}
+	}
+
+	return nil
+}
+
+// newMapClaimSessionFn returns a new MapFn that creates a claim for the given
+// session. Any session which encouters an error while creating a claim is sent
+// on the failedCreateClaimSessions channel.
+func (mnr *miner) newMapClaimSessionFn(
+	failedCreateClaimSessions chan<- relayer.SessionTree,
+) channel.MapFn[relayer.SessionTree, either.SessionTree] {
+	return func(
+		ctx context.Context,
+		session relayer.SessionTree,
+	) (_ either.SessionTree, skip bool) {
+		// this session should no longer be updated
+		claimRoot, err := session.Flush()
+		if err != nil {
+			return either.Error[relayer.SessionTree](err), false
+		}
+
+		sessionHeader := session.GetSessionHeader()
+		if err := mnr.supplierClient.CreateClaim(ctx, *sessionHeader, claimRoot); err != nil {
+			failedCreateClaimSessions <- session
+			return either.Error[relayer.SessionTree](err), false
+		}
+
+		return either.Success(session), false
+	}
+}
+
+// mapWaitForOpenProofWindow maps over the claimedSessions observable, applying
+// the mapWaitForOpenProofWindow method to each session. It returns an observable
```

Comment:
This comment is a bit confusing. 

1. We do not return an observable from this function 
2. We use `waitForEarliestSubmitProofDistributionHeight` instead of `mapWaitForOpenProofWindow`

Can you clarify what's intended here?